### DISCUSSION TITLE
anonymous struct ( declaring struct in a function )

### DISCUSSION THREAD

**Discussion by ghost** [vlang/v#14671] [Ideas]
Nice to have an anonymous `struct` that can be disposable in a function.
It is a struct that is never used outside of that function.
For now, I am dealing with multidimensional arrays of strings whenever possible.
It would be useful if it could be used in arrays with the type as it is without conversion to string

**Comment by medvednikov** [comment]
yeah we'll have that

**Comment by medvednikov** [comment]
well i was referring to

```
struct Foo {
  bar struct { 
    x int
  }
}
```

**Comment by ghost** [comment]
Sorry I didn't make it clear enough. 
Perhaps it would have been better to say that a struct can be declared within a function, like an anonymous function.
```v
fn foo(){
    mut res := []Res{
        a string
        n int
    }
    for .... {
        res << ....
    }
}
```
I want to confine it to a function because it is only used within that function and I use all the same names like Res or Item every time.

It would be useful for making a list from sql result or returned multiple-values.

**Comment by ghost** [comment]
Oh, but this is very nice.
```v
fn foo(){
    mut res := []struct{
        a string
        n int
        x []struct{
            b int
        }
    }
    ....
}
```

**Comment by Wajinn** [comment]
> ...Perhaps it would have been better to say that a struct can be declared within a function, like an anonymous function.
> I want to confine it to a function because it is only used within that function...
> ```v
> fn foo(){
>     mut res := []Res{
>         a string
>         n int
>     }
>     for .... {
>         res << ....
>     }
> }
> ```
I like Kahsa's idea of being able to confine a struct to a particular function.  There are cases where a struct is only used for a particular function, and is unnecessary beyond that.  Though when it is declared, would rather keep to using the keyword `struct`, so that the type is more obvious.  

However, I'm not sure about the full ramifications of such changes.  Maybe there needs to be some debate among the contributors and developers as to the plus versus minus of it.

**Comment by medvednikov** [comment]
We'll have structs inside functions too. But like in Go, it won't be possible to declare methods on them.

**Reply by Wajinn** [reply]
> We'll have structs inside functions too. But like in Go, it won't be possible to declare methods on them.

Nice.  I could live with such restrictions, for the additional usefulness.

**Comment by chipnetics** [comment]
Just chiming in, this is a feature I very much look forward to from a "code organization" perspective. I have a lot of one-off structs that sit outside functions and feel wrong!

**Comment by ghost** [comment]
It doesn't seem to be included in the 0.4 roadmap, just forget to include it?

If possible, I would like to see a list of additional features currently planned.
It would also be better if we knew what version is scheduled to be implemented.

**Reply by medvednikov** [reply]
It's a minor feature that's fast to implement, the roadmap only has big projects. I'll add it anyway.

**Comment by gprst** [comment]
Up 😊 

This feature prevents for instance having to write a `GetSomePayloadResponse` struct for every single function that makes a `GET` HTTP call. If it is fast and easy to implement, do you think it could be done for the 0.3.4 release? @medvednikov 

I am very sorry to be the one that asks the usual "Any update on this?" post, but this feature seems to have been forgotten for a year. I would absolutely understand if this was so low priority it would be deferred to some years in the future 😊

**Reply by JalonSolov** [reply]
0.3.4 was released on April 30th.  It would have to be the 0.3.5 release, at the earliest.

**Reply by gprst** [reply]
Yes, that's what I meant 😊
